controlpanel/api/tools.py

```diff
@@ -0,0 +1,237 @@
+from collections import UserDict
+import secrets
+
+from django.conf import settings
+from django.utils.functional import cached_property
+from rest_framework.exceptions import APIException
+
+from controlpanel.api.helm import Helm
+from controlpanel.kubeapi.views import client
+
+
+IDLED = "mojanalytics.xyz/idled"
+
+SUPPORTED_TOOL_NAMES = list(settings.TOOLS.keys())
+
+
+class DuplicateTool(Exception):
+    pass
+
+
+class ToolNotDeployed(APIException):
+    status_code = 400
+    default_detail = f"Specified tool not deployed"
+    default_code = "tool_not_deployed"
+
+
+class UnsupportedToolException(APIException):
+    status_code = 400
+    supported = ", ".join(SUPPORTED_TOOL_NAMES)
+    default_detail = f"Unsupported tool, must be one of: {supported}"
+    default_code = "unsupported_tool"
+
+
+class ToolRepository(UserDict):
+
+    def __getitem__(self, key):
+        if key not in self:
+            raise UnsupportedToolException(key)
+
+        if key not in self:
```

Comment:
Isn't this `if` statement duplicated? Maybe you wanted something like `if key not in SUPPORTED_TOOL_NAMES` here as well? Not sure.

Reply:
This is a mistake - not sure what I meant to do here, but this duplication is wrong.